### PR TITLE
feat(protocol-designer): implement step selection banner

### LINKED ISSUE #7081: Step Selection Banner with Step Type & Count
**User Story**
As a PD user, I would like to be able to see how many and what types of steps I have selected for editing, so that I can be assured I am editing the correct batch of steps

**Description**
This ticket accounts for the step selection banner, step count, and step type pill outlines. The step selection banner is displayed as the blue box above the step form, as seen below

**Requirements**

1. Step selection banner appears at top of form when multi-select mode is entered.
2. Step selection banner contains count of steps selected. The count changes dynamically as steps are selected and deselected. It appears in two places in the banner (see screen above):
* As part of a headline string: ("[count] Steps Selected")
* As part of the step type button outline 
3. Step selection banner contains step type box outline, which shows the user the types of steps they have selected, as well as an icon and count for those steps (see screen above). Step types should change dynamically as a user selects / deselects steps
4. Step type wraps as pills add up, and the 'exit batch edit' button stays at the top right of the banner

**Acceptance Criteria**
I've created a template protocol for this test, ask me (AW) for it when the time comes

1. Create a protocol with 3 transfer steps, 2 mix steps, and 1 each of other steps
2. Select a transfer step
3. Hold shift to enter multi-select mode. Banner should appear at top of form area, displaying count "[1] Step Selected" and step type (Transfer) 
4. Add two transfer steps. Banner should display count "[3] Steps Selected" and step type (Transfer [3])
5. Add two mix steps. Banner should display count "[5] Steps Selected" and step types (Transfer [3]) (Mix [2]) 
6. Add one of each other step. Banner should display count "[9] Steps Selected" and step types (Transfer [3]) (Mix [2]) (Pause [1]) (Magnet [1]) (Temperature [1]) (Thermocycler [1]) 
7. Remove one transfer step. Banner should update accordingly "[8] Steps Selected", Transfer [2])
8. Remove one mix step. Banner should update accordingly "[7] Steps Selected", Mix [1])
9. Remove each remaining step type, including both transfer steps. Banner should updated accordingly (step count reduced, step type removed)

### PR DESCRIPTION
**Pull request by IanLondon**
# Overview

Closes #7081

# Changelog

- Create step selection banner and hook it up
- add new icon `checkbox-multiple-marked-outline`

# Review requests

Dispatch an action via Redux dev tools like:

```js
{
type: 'SELECT_MULTIPLE_STEPS',
    payload: {
      stepIds: ["05d39a20-4143-11eb-956a-7fedb1f0450f", "2dd685a0-4143-11eb-956a-7fedb1f0450f"],
      lastSelected: 'whatever' 
    },
}
```

You need the `payload.stepIds` to actually match your stepIds, look at `orderedStepIds` in your state.

Then, you should see the multiselect banner with the pills for each step type containing the count of each step etc as described thoroughly in the ticket (including responsiveness: pills should stack)

NOTE: the "exit multiselect mode" button doesn't do anything. I figure we'll do that when we implement "exit multiselect mode" in the control bar, it ought to be made to work exactly the same.

# Risk assessment

Low, should be effectively under a FF (the banner isn't directly hidden under a FF, but multiselect mode in general is, so a user has no path to making this banner appear until multiselection behavior is added to PD)

### CODE REVIEW

File: protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    i18n.t(`application.stepType.${stepType}`)
+  )} (${count})`
+  return (
+    <Flex
+      alignItems={ALIGN_CENTER}
+      border={`${BORDER_WIDTH_DEFAULT} ${BORDER_STYLE_SOLID} ${C_SELECTED_DARK}`}
+      borderRadius="20px"
+      padding="0.5rem"
+      marginRight="0.5rem"
+      marginBottom="0.5rem"
```
Comment (shlokamin) [inline protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js:48]: Suggestion for the future — it would be cool to start using agreed upon spacial convention for things like padding and margin. Design could then say things like "this component should use standard border radius and small margin" or something.

Comment (shlokamin) [reply protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js:48]: @howisthisnamenottakenyet

File: protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+export const StepSelectionBannerComponent = (props: Props): React.Node => {
+  const { selectedSteps, handleExitBatchEdit } = props
+  const numSteps = selectedSteps.length
+  const countPerType = selectedSteps.reduce((acc, step) => {
+    const { stepType } = step
+    const newCount = acc[stepType] ? acc[stepType] + 1 : 1
+    return { ...acc, [stepType]: newCount }
+  }, {})
+  const stepTypes = Object.keys(countPerType).sort()
```
Comment (shlokamin) [inline protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js:67]: will this work for more than 9 steps (when string representation gets more complicated)? I know sort can be finicky if you don't explicitly pass in a comparison function. The default sort function will sort by converting elements to string and comparing against their unicode value.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

File: protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      justifyContent={JUSTIFY_SPACE_BETWEEN}
+    >
+      <Box flex="0 1 auto">
+        <Flex alignItems={ALIGN_CENTER}>
+          <Icon name="checkbox-multiple-marked-outline" width={SIZE_2} />
+          <Text
+            width="10rem"
+            marginLeft="0.5rem"
+            fontWeight={FONT_WEIGHT_SEMIBOLD}
+            textTransform={TEXT_TRANSFORM_UPPERCASE}
```
Comment (shlokamin) [inline protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js:83]: Ahhh this is what I was talking about in my comment above, but for space instead of font

File: protocol-designer/src/localization/en/application.json [hunk 4]
```diff
@@ -27,5 +27,7 @@
     "generic_verification_failure": "Something went wrong with your unique link. Fill out the form below to have a new one sent to your email. Please contact the Opentrons Support team if you require further help.",
     "unauthorized_verification_failure": "This unique link has expired and is no longer valid, to have a new link sent to your email, fill out the form below.",
     "validation_server_failure": "There was a problem communicating with the Opentrons validation server. Please make sure you are still connected to the internet and try again. If this problem persists, please contact Opentrons support."
-  }
+  },
+  "exit_batch_edit": "exit batch edit",
+  "n_steps_selected": "{{n}} steps selected"
```
Comment (shlokamin) [inline protocol-designer/src/localization/en/application.json:32]: yay for `i18n`!

File: protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js (hunk 2 above)
Comment (shlokamin) [reply protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js:67]:
```js
[1,3,44,111,12222].sort() // => [1, 111, 12222, 3, 44]
```

**Review by shlokamin — COMMENTED** [review]
Looks good! Left a comment about sorting but thats it

File: protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js (hunk 2 above)
Comment (IanLondon) [reply protocol-designer/src/components/StepSelectionBanner/StepSelectionBannerComponent.js:67]: it sorts the keys, which are `StepType` not the count. It's a somewhat meaningless sort, but I figured it was better than the raw output of `Object.keys`, which has an undefined ordering. At least `Object.keys(blah).sort()` is defined. AFAIK, the AC don't say what order the steps should be. If we want strict alpha order we would need to deal with how StepType doesn't match the display name, but I just ignored that.

tl;dr I'm not really sorting meaningfully here I'm just avoiding potential (and unlikely) variation that could come from displaying the output of `Object.keys`'s undefined ordering

**Review by shlokamin — APPROVED** [review]

**Review by Kadee80 — APPROVED** [review]
🍌
